### PR TITLE
module_helper cmd - added feature flag to control whether CmdMixin adds rc, out and err t…

### PR DESCRIPTION
**Pull request by russoz**
##### SUMMARY
Added feature flag to control whether `CmdMixin` adds `rc`, `out` and `err` to the result of the module.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
plugins/module_utils/mh/mixins/cmd.py

### CODE REVIEW

File: plugins/module_utils/mh/mixins/cmd.py [hunk 1]
```diff
@@ -88,6 +88,7 @@ class CmdMixin(object):
     """
     command = None
     command_args_formats = {}
+    command_output_as_result = True
```
Comment (felixfontein) [inline plugins/module_utils/mh/mixins/cmd.py:91]: I don't think this should be a global option, but (also) a parameter to `run_command`.

Also there are cases (such as for `snap`) where being able to just not add `rc` could be useful. Having `stdout` and `stderr` from the last executed command can be useful for troubleshooting.

Comment (russoz) [reply plugins/module_utils/mh/mixins/cmd.py:91]: Yeah, it makes more sense to make it per-call to run_command.

**Comment by russoz** [comment]
Not sure I like the term `publish` in there, and I am thinking that maybe I should use just one single parameter and bit mask the options for rc, out and err. But that will have to wait for tomorrow.

File: plugins/module_utils/mh/mixins/cmd.py [hunk 2]
```diff
@@ -166,7 +173,12 @@ def run_command(self, extra_params=None, params=None, process_output=None, *args
             self.update_output(force_lang=self.force_lang)
             options['environ_update'] = env_update
         rc, out, err = self.module.run_command(self.vars.cmd_args, *args, **options)
-        self.update_output(rc=rc, stdout=out, stderr=err)
+        if publish_rc:
+            self.update_output(rc=rc, stdout=out, stderr=err)
+        if publish_out:
+            self.update_output(rc=rc, stdout=out, stderr=err)
+        if publish_err:
+            self.update_output(rc=rc, stdout=out, stderr=err)
```
Comment (felixfontein) [inline plugins/module_utils/mh/mixins/cmd.py:181]: So what's the difference between `publish_rc`, `publish_out` and `publish_err`? I think you need to adjust the `update_output` calls a bit ;)

Comment (russoz) [reply plugins/module_utils/mh/mixins/cmd.py:181]: Yeah, it was that late at night. Wrapping this up today.

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
Didn't we leave the "good old times" when we had to do bit operations for such things behind? :)

**Comment by russoz** [comment]
I am from the old guard. Should I revert back to separate params? I like the bitwise way because it makes it simple to express "all" and "none". With the 3-params we need to set each one individually.

**Comment by felixfontein** [comment]
That's a very good question. Let's ask some more people on IRC over the next days :)

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@russoz if you want to merge this as-is, just press the button. If you want to add more changes or discuss this, feel free to do so :)

**Comment by russoz** [comment]
force-pushed after a local rebase

**Comment by russoz** [comment]
I'll go for it as-is.

**Pull request merged**